File: code.py
#!/usr/bin/env python3
import sys

class Term:
    """项：表示 a * x^b * sin^c(x) * cos^d(x)"""
    __slots__ = ('a', 'b', 'c', 'd')

    def __init__(self, a: int = 0, b: int = 0, c: int = 0, d: int = 0):
        self.a = a
        self.b = b
        self.c = c
        self.d = d

    def __eq__(self, other) -> bool:
        return self.b == other.b and self.c == other.c and self.d == other.d

    def __ne__(self, other) -> bool:
        return not self.__eq__(other)


class Poly:
    """多项式：由若干 Term 组成"""
    __slots__ = ('terms',)

    def __init__(self, size_or_terms=None):
        if isinstance(size_or_terms, int):
            self.terms = [Term() for _ in range(size_or_terms)]
        elif isinstance(size_or_terms, list):
            self.terms = size_or_terms[:]
        else:
            self.terms = []

    def __len__(self) -> int:
        return len(self.terms)

    def simplify(self) -> None:
        """合并同类项"""
        self.terms.sort(key=lambda t: (t.b, t.c, t.d), reverse=True)

        if not self.terms:
            return

        merged = []
        i = 0
        while i < len(self.terms):
            current = Term(self.terms[i].a, self.terms[i].b, self.terms[i].c, self.terms[i].d)
            j = i + 1
            while j < len(self.terms) and self.terms[j] == current:
                current.a += self.terms[j].a
                j += 1
            if current.a != 0:
                merged.append(current)
            i = j

        self.terms = merged

    def __add__(self, other: 'Poly') -> 'Poly':
        result_terms = [Term(t.a, t.b, t.c, t.d) for t in self.terms]
        result_terms.extend([Term(t.a, t.b, t.c, t.d) for t in other.terms])
        result = Poly(result_terms)
        result.simplify()
        return result

    def __sub__(self, other: 'Poly') -> 'Poly':
        result_terms = [Term(t.a, t.b, t.c, t.d) for t in self.terms]
        result_terms.extend([Term(-t.a, t.b, t.c, t.d) for t in other.terms])
        result = Poly(result_terms)
        result.simplify()
        return result

    def __mul__(self, other: 'Poly') -> 'Poly':
        result_terms = []
        for t1 in self.terms:
            for t2 in other.terms:
                new_term = Term(
                    t1.a * t2.a,
                    t1.b + t2.b,
                    t1.c + t2.c,
                    t1.d + t2.d
                )
                result_terms.append(new_term)

        result = Poly(result_terms)
        result.simplify()
        return result

    def __copy__(self) -> 'Poly':
        return Poly([Term(t.a, t.b, t.c, t.d) for t in self.terms])

    def derivate(self) -> 'Poly':
        """对多项式求导"""
        result_terms = []

        for t in self.terms:
            if t.b > 0:
                result_terms.append(Term(t.a * t.b, t.b - 1, t.c, t.d))
            if t.c > 0:
                result_terms.append(Term(t.a * t.c, t.b, t.c - 1, t.d + 1))
            if t.d > 0:
                result_terms.append(Term(-t.a * t.d, t.b, t.c + 1, t.d - 1))

        result = Poly(result_terms)
        result.simplify()
        return result


class Frac:
    """分式：p / q"""
    __slots__ = ('p', 'q')

    def __init__(self, arg1=None, arg2=None):
        if arg1 is None:
            self.p = Poly()
            self.q = Poly()
        elif isinstance(arg1, int):
            self.p = Poly([Term(arg1, 0, 0, 0)])
            self.q = Poly([Term(1, 0, 0, 0)])
        elif isinstance(arg1, Term):
            self.p = Poly([Term(arg1.a, arg1.b, arg1.c, arg1.d)])
            self.q = Poly([Term(1, 0, 0, 0)])
        elif isinstance(arg1, Poly) and isinstance(arg2, Poly):
            self.p = arg1.__copy__()
            self.q = arg2.__copy__()
        else:
            raise TypeError("Invalid Frac constructor arguments")

    def __add__(self, other: 'Frac') -> 'Frac':
        new_p = self.p * other.q + other.p * self.q
        new_q = self.q * other.q
        return Frac(new_p, new_q)

    def __sub__(self, other: 'Frac') -> 'Frac':
        new_p = self.p * other.q - other.p * self.q
        new_q = self.q * other.q
        return Frac(new_p, new_q)

    def __mul__(self, other: 'Frac') -> 'Frac':
        new_p = self.p * other.p
        new_q = self.q * other.q
        return Frac(new_p, new_q)

    def __truediv__(self, other: 'Frac') -> 'Frac':
        new_p = self.p * other.q
        new_q = self.q * other.p
        return Frac(new_p, new_q)

    def derivate(self) -> 'Frac':
        p_prime = self.p.derivate()
        q_prime = self.q.derivate()
        new_p = p_prime * self.q - q_prime * self.p
        new_q = self.q * self.q
        return Frac(new_p, new_q)

    def output(self) -> str:
        def format_term(t: Term) -> str:
            parts = []

            if t.b == 0 and t.c == 0 and t.d == 0:
                return str(abs(t.a))
            elif abs(t.a) != 1:
                parts.append(str(abs(t.a)))

            if t.b > 0:
                if t.b == 1:
                    parts.append('x')
                else:
                    parts.append(f'x^{t.b}')

            if t.c > 0:
                if t.c == 1:
                    parts.append('sinx')
                else:
                    parts.append(f'sin^{t.c}x')

            if t.d > 0:
                if t.d == 1:
                    parts.append('cosx')
                else:
                    parts.append(f'cos^{t.d}x')

            return ''.join(parts)

        def format_poly(poly: Poly) -> str:
            if not poly.terms:
                return '0'

            result = []
            for i, t in enumerate(poly.terms):
                term_str = format_term(t)
                if i == 0:
                    if t.a < 0:
                        result.append('-' + term_str)
                    else:
                        result.append(term_str)
                else:
                    if t.a < 0:
                        result.append('-' + term_str)
                    else:
                        result.append('+' + term_str)

            return ''.join(result)

        if not self.p.terms:
            return '0'

        p_str = format_poly(self.p)

        if len(self.q.terms) == 1 and self.q.terms[0].a == 1 and \
           self.q.terms[0].b == 0 and self.q.terms[0].c == 0 and self.q.terms[0].d == 0:
            return p_str

        q_str = format_poly(self.q)

        if len(self.p.terms) > 1:
            p_str = f'({p_str})'
        if len(self.q.terms) > 1:
            q_str = f'({q_str})'

        return f'{p_str}/{q_str}'


def parse_term(s: str, start: int, end: int) -> Term:
    """解析一个项 s[start:end]"""
    i = start

    # 解析系数
    coeff = 1
    neg = False

    if i < end and s[i] == '-':
        neg = True
        i += 1
    elif i < end and s[i] == '+':
        i += 1

    # 读取数字系数
    num_start = i
    while i < end and s[i].isdigit():
        i += 1

    if i > num_start:
        coeff = int(s[num_start:i])

    if neg:
        coeff = -coeff

    # 解析 x, sinx, cosx
    b, c, d = 0, 0, 0

    while i < end:
        if i + 4 <= end and s[i:i+4] == 'sinx':
            i += 4
            if i < end and s[i] == '^':
                i += 1
                exp_start = i
                while i < end and s[i].isdigit():
                    i += 1
                c = int(s[exp_start:i])
            else:
                c = 1
        elif i + 4 <= end and s[i:i+4] == 'cosx':
            i += 4
            if i < end and s[i] == '^':
                i += 1
                exp_start = i
                while i < end and s[i].isdigit():
                    i += 1
                d = int(s[exp_start:i])
            else:
                d = 1
        elif s[i] == 'x':
            i += 1
            if i < end and s[i] == '^':
                i += 1
                exp_start = i
                while i < end and s[i].isdigit():
                    i += 1
                b = int(s[exp_start:i])
            else:
                b = 1
        else:
            i += 1

    # 如果没有任何变量，系数部分已经包含全部
    if b == 0 and c == 0 and d == 0 and num_start == start:
        # 这是纯常数
        pass

    return Term(coeff, b, c, d)


def parse_poly(s: str, l: int, r: int) -> Poly:
    """解析多项式（加减项）"""
    terms = []
    i = l

    while i < r:
        # 找下一个加减号（不在括号内）
        j = i
        if i < r and s[i] in '+-':
            j = i + 1
        else:
            j = i

        # 向后查找，直到遇到下一个 +- 或结束
        while j < r and s[j] not in '+-':
            j += 1

        # 解析 s[i:j]
        if i < j:
            term = parse_term(s, i, j)
            terms.append(term)

        i = j

    poly = Poly(terms)
    poly.simplify()
    return poly


def find_paren(s: str, pos: int) -> int:
    """找到与 pos 处左括号匹配的右括号"""
    depth = 1
    i = pos + 1
    while i < len(s):
        if s[i] == '(':
            depth += 1
        elif s[i] == ')':
            depth -= 1
            if depth == 0:
                return i
        i += 1
    return len(s) - 1


def find_op(s: str, l: int, r: int, ops: list) -> int:
    """从右向左找第一个不在括号内的运算符"""
    depth = 0
    for i in range(r - 1, l - 1, -1):
        if s[i] == ')':
            depth += 1
        elif s[i] == '(':
            depth -= 1
        elif depth == 0 and s[i] in ops:
            return i
    return -1


def dfs(s: str, l: int, r: int) -> Frac:
    """递归下降解析"""
    # 去掉外层括号
    while l < r and s[l] == '(' and find_paren(s, l) == r - 1:
        l += 1
        r -= 1

    # 找 + -
    pos = find_op(s, l, r, ['+', '-'])
    if pos != -1:
        left = dfs(s, l, pos)
        right = dfs(s, pos + 1, r)
        return left + right if s[pos] == '+' else left - right

    # 找 * /
    pos = find_op(s, l, r, ['*', '/'])
    if pos != -1:
        left = dfs(s, l, pos)
        right = dfs(s, pos + 1, r)
        return left * right if s[pos] == '*' else left / right

    # 基础情况
    poly = parse_poly(s, l, r)
    return Frac(poly, Poly([Term(1, 0, 0, 0)]))


def solve(s: str) -> None:
    n = len(s)
    frac = dfs(s, 0, n)
    print(frac.output())

    frac_prime = frac.derivate()
    print(frac_prime.output())


def main():
    expr = sys.stdin.readline().strip()
    solve(expr)


if __name__ == "__main__":
    main()

File: code.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <cstring>
using namespace std;

struct Term {
    int a, b, c, d;  // a * x^b * sin^c(x) * cos^d(x)

    Term(int _a = 0, int _b = 0, int _c = 0, int _d = 0) : a(_a), b(_b), c(_c), d(_d) {}

    bool operator==(const Term& other) const {
        return b == other.b && c == other.c && d == other.d;
    }

    bool operator<(const Term& other) const {
        if (b != other.b) return b > other.b;
        if (c != other.c) return c > other.c;
        return d > other.d;
    }
};

struct Poly {
    vector<Term> terms;

    void simplify() {
        sort(terms.begin(), terms.end());
        vector<Term> result;

        for (int i = 0; i < terms.size(); ) {
            Term cur = terms[i];
            int j = i + 1;
            while (j < terms.size() && terms[j] == cur) {
                cur.a += terms[j].a;
                j++;
            }
            if (cur.a != 0) {
                result.push_back(cur);
            }
            i = j;
        }

        terms = result;
    }

    Poly operator+(const Poly& other) const {
        Poly result;
        result.terms = terms;
        result.terms.insert(result.terms.end(), other.terms.begin(), other.terms.end());
        result.simplify();
        return result;
    }

    Poly operator-(const Poly& other) const {
        Poly result;
        result.terms = terms;
        for (const auto& t : other.terms) {
            result.terms.push_back(Term(-t.a, t.b, t.c, t.d));
        }
        result.simplify();
        return result;
    }

    Poly operator*(const Poly& other) const {
        Poly result;
        for (const auto& t1 : terms) {
            for (const auto& t2 : other.terms) {
                result.terms.push_back(Term(t1.a * t2.a, t1.b + t2.b, t1.c + t2.c, t1.d + t2.d));
            }
        }
        result.simplify();
        return result;
    }

    Poly derivate() const {
        Poly result;
        for (const auto& t : terms) {
            if (t.b > 0) {
                result.terms.push_back(Term(t.a * t.b, t.b - 1, t.c, t.d));
            }
            if (t.c > 0) {
                result.terms.push_back(Term(t.a * t.c, t.b, t.c - 1, t.d + 1));
            }
            if (t.d > 0) {
                result.terms.push_back(Term(-t.a * t.d, t.b, t.c + 1, t.d - 1));
            }
        }
        result.simplify();
        return result;
    }
};

struct Frac {
    Poly p, q;

    Frac() {}
    Frac(const Poly& _p, const Poly& _q) : p(_p), q(_q) {}

    Frac operator+(const Frac& other) const {
        Frac result;
        result.p = p * other.q + other.p * q;
        result.q = q * other.q;
        return result;
    }

    Frac operator-(const Frac& other) const {
        Frac result;
        result.p = p * other.q - other.p * q;
        result.q = q * other.q;
        return result;
    }

    Frac operator*(const Frac& other) const {
        Frac result;
        result.p = p * other.p;
        result.q = q * other.q;
        return result;
    }

    Frac operator/(const Frac& other) const {
        Frac result;
        result.p = p * other.q;
        result.q = q * other.p;
        return result;
    }

    Frac derivate() const {
        Frac result;
        Poly p_prime = p.derivate();
        Poly q_prime = q.derivate();
        result.p = p_prime * q - q_prime * p;
        result.q = q * q;
        return result;
    }

    void output() const {
        auto format_term = [](const Term& t) -> string {
            if (t.b == 0 && t.c == 0 && t.d == 0) {
                return to_string(abs(t.a));
            }

            string result;
            if (abs(t.a) != 1) {
                result += to_string(abs(t.a));
            }

            if (t.b > 0) {
                result += "x";
                if (t.b > 1) {
                    result += "^" + to_string(t.b);
                }
            }

            if (t.c > 0) {
                if (t.c > 1) {
                    result += "sin^" + to_string(t.c);
                } else {
                    result += "sin";
                }
                result += "x";
            }

            if (t.d > 0) {
                if (t.d > 1) {
                    result += "cos^" + to_string(t.d);
                } else {
                    result += "cos";
                }
                result += "x";
            }

            return result;
        };

        auto format_poly = [&](const Poly& poly) -> string {
            if (poly.terms.empty()) {
                return "0";
            }

            string result;
            for (int i = 0; i < poly.terms.size(); i++) {
                const auto& t = poly.terms[i];
                string term_str = format_term(t);

                if (i == 0) {
                    if (t.a < 0) {
                        result += "-" + term_str;
                    } else {
                        result += term_str;
                    }
                } else {
                    if (t.a < 0) {
                        result += "-" + term_str;
                    } else {
                        result += "+" + term_str;
                    }
                }
            }

            return result;
        };

        if (p.terms.empty()) {
            cout << "0\n";
            return;
        }

        string p_str = format_poly(p);

        if (q.terms.size() == 1 && q.terms[0].a == 1 &&
            q.terms[0].b == 0 && q.terms[0].c == 0 && q.terms[0].d == 0) {
            cout << p_str << "\n";
            return;
        }

        string q_str = format_poly(q);

        if (p.terms.size() > 1) {
            p_str = "(" + p_str + ")";
        }
        if (q.terms.size() > 1) {
            q_str = "(" + q_str + ")";
        }

        cout << p_str << "/" << q_str << "\n";
    }
};

string s;

int get_number(int l, int r) {
    if (l >= r) return 1;

    bool neg = false;
    int start = l;

    if (s[l] == '-') {
        neg = true;
        start = l + 1;
    } else if (s[l] == '+') {
        start = l + 1;
    }

    int num = 0;
    bool hasDigit = false;
    for (int i = start; i < r && isdigit(s[i]); i++) {
        num = num * 10 + (s[i] - '0');
        hasDigit = true;
    }

    if (!hasDigit) return neg ? -1 : 1;
    return neg ? -num : num;
}

Term get_term(int l, int r) {
    int i = l;

    // Parse coefficient
    int start = i;
    if (i < r && (s[i] == '-' || s[i] == '+')) {
        i++;
    }
    while (i < r && isdigit(s[i])) {
        i++;
    }

    int coeff = get_number(start, i);

    // Parse x, sin, cos
    int b = 0, c = 0, d = 0;

    while (i < r) {
        if (i + 3 <= r && s.substr(i, 3) == "sin") {
            i += 3;
            if (i < r && s[i] == '^') {
                i++;
                int exp_start = i;
                while (i < r && isdigit(s[i])) {
                    i++;
                }
                c = stoi(s.substr(exp_start, i - exp_start));
            } else {
                c = 1;
            }
            if (i < r && s[i] == 'x') {
                i++;
            }
        } else if (i + 3 <= r && s.substr(i, 3) == "cos") {
            i += 3;
            if (i < r && s[i] == '^') {
                i++;
                int exp_start = i;
                while (i < r && isdigit(s[i])) {
                    i++;
                }
                d = stoi(s.substr(exp_start, i - exp_start));
            } else {
                d = 1;
            }
            if (i < r && s[i] == 'x') {
                i++;
            }
        } else if (s[i] == 'x') {
            i++;
            if (i < r && s[i] == '^') {
                i++;
                int exp_start = i;
                while (i < r && isdigit(s[i])) {
                    i++;
                }
                b = stoi(s.substr(exp_start, i - exp_start));
            } else {
                b = 1;
            }
        } else {
            i++;
        }
    }

    return Term(coeff, b, c, d);
}

Poly parse_poly(int l, int r) {
    Poly result;

    int i = l;
    while (i < r) {
        int j = i;
        if (i < r && (s[i] == '+' || s[i] == '-')) {
            j = i + 1;
        } else {
            j = i;
        }

        while (j < r && s[j] != '+' && s[j] != '-') {
            j++;
        }

        if (i < j) {
            result.terms.push_back(get_term(i, j));
        }

        i = j;
    }

    result.simplify();
    return result;
}

int find_paren(int pos) {
    int depth = 1;
    int i = pos + 1;
    while (i < s.length()) {
        if (s[i] == '(') depth++;
        else if (s[i] == ')') {
            depth--;
            if (depth == 0) return i;
        }
        i++;
    }
    return s.length() - 1;
}

int find_op(int l, int r, const string& ops) {
    int depth = 0;
    for (int i = r - 1; i >= l; i--) {
        if (s[i] == ')') depth++;
        else if (s[i] == '(') depth--;
        else if (depth == 0 && ops.find(s[i]) != string::npos) {
            return i;
        }
    }
    return -1;
}

Frac dfs(int l, int r) {
    // Remove outer parentheses
    while (l < r && s[l] == '(' && find_paren(l) == r - 1) {
        l++;
        r--;
    }

    // Find + or -
    int pos = find_op(l, r, "+-");
    if (pos != -1) {
        Frac left = dfs(l, pos);
        Frac right = dfs(pos + 1, r);
        return s[pos] == '+' ? left + right : left - right;
    }

    // Find * or /
    pos = find_op(l, r, "*/");
    if (pos != -1) {
        Frac left = dfs(l, pos);
        Frac right = dfs(pos + 1, r);
        return s[pos] == '*' ? left * right : left / right;
    }

    // Base case
    Poly p = parse_poly(l, r);
    Poly q;
    q.terms.push_back(Term(1, 0, 0, 0));
    return Frac(p, q);
}

void solve() {
    Frac frac = dfs(0, s.length());
    frac.output();

    Frac frac_prime = frac.derivate();
    frac_prime.output();
}

int main() {
    getline(cin, s);
    solve();
    return 0;
}
